include/ghex/structured/cubed_sphere/field_descriptor.hpp

```diff
@@ -0,0 +1,504 @@
+                                return -(*reinterpret_cast<const T*>(reinterpret_cast<const char*>(m_buffer_desc.m_ptr) +
+                                    memory_location));
+                            // reverse the y component if necessary
+                            else if (m_is_vector && coord[0]==1 && m_transform.m_rotation[1]==-1)
+                                return -(*reinterpret_cast<const T*>(reinterpret_cast<const char*>(m_buffer_desc.m_ptr) +
+                                    memory_location));
+                            else
+                                return *reinterpret_cast<const T*>(reinterpret_cast<const char*>(m_buffer_desc.m_ptr) +
+                                    memory_location);
+                        }
+
+                        /** @brief accesses field at specified local coordinate
+                          * @param coord 4-dimensional array (component, x, y, z) in local coordinate system
+                          * @return reference to the value in the field */
+                        GT_FUNCTION
+                        T& data(const coordinate_type& coord) const noexcept {
+                            // make data memory coordinates from local coordinates
+                            const coordinate_type data_coord = coord + m_data_is.m_offset;
+                            // dot product with strides to compute address
+                            const auto memory_location =
+                                m_data_is.m_strides[0]*data_coord[0] +
+                                m_data_is.m_strides[1]*data_coord[1] +
+                                m_data_is.m_strides[2]*data_coord[2] +
+                                m_data_is.m_strides[3]*data_coord[3];
+                            return *reinterpret_cast<T*>(
+                                reinterpret_cast<char*>(m_data_is.m_ptr) + memory_location);
+                        }
+                    };
+
+                private: // members
+                    domain_id_type  m_dom_id;        /// domain id
+                    int             m_c;             /// cube size
+                    bool            m_is_vector;     /// is this data describing a vector field 
+                    value_type*     m_data;          /// pointer to data
+                    coordinate_type m_dom_first;     /// global coordinate of first domain cell
+                    coordinate_type m_offsets;       /// offset from beginning of memory to the first domain cell
+                    coordinate_type m_extents;       /// extent of memory (including halos)
+                    device_id_type  m_device_id;     /// device id
+                    strides_type    m_byte_strides;  /// memory strides in bytes
+
+                public: // ctors
+                    /** @brief construct field_descriptor from existing data
+                      * @tparam Array 3-dimensional array-like type
```

Comment:
Is it 3D only?

Reply:
the domain id 3D always, the data can be 4D